File: backend/internal/services/course.go
package services

import (
	"errors"
	"fmt"

	"github.com/imlargo/go-api-template/internal/dto"
	"github.com/imlargo/go-api-template/internal/models"
	"github.com/imlargo/go-api-template/pkg/utils"
)

type CourseService interface {
	CreateCourse(course *models.Course) (*models.Course, error)
	GetCourse(id uint) (*models.Course, error)
	UpdateCourse(id uint, course *models.Course) (*models.Course, error)
	UpdateCoursePatch(id uint, data map[string]interface{}) (*models.Course, error)
	DeleteCourse(id uint) error
	GetAllCourses() ([]*models.Course, error)
	GetCourseWithModules(id uint) (*models.Course, error)
	GetCoursesWithEnrollmentCount() ([]*models.Course, error)
}

type courseService struct {
	*Service
}

func NewCourseService(service *Service) CourseService {
	return &courseService{
		Service: service,
	}
}

func (s *courseService) CreateCourse(course *models.Course) (*models.Course, error) {
	if err := s.store.Courses.Create(course); err != nil {
		return nil, fmt.Errorf("error al crear el curso: %w", err)
	}
	return course, nil
}

func (s *courseService) GetCourse(id uint) (*models.Course, error) {
	course, err := s.store.Courses.Get(id)
	if err != nil {
		return nil, fmt.Errorf("curso no encontrado: %w", err)
	}
	return course, nil
}

func (s *courseService) UpdateCourse(id uint, courseData *models.Course) (*models.Course, error) {
	existingCourse, err := s.store.Courses.Get(id)
	if err != nil {
		return nil, fmt.Errorf("curso no encontrado: %w", err)
	}

	// Update fields
	existingCourse.Title = courseData.Title
	existingCourse.Description = courseData.Description
	existingCourse.ShortDescription = courseData.ShortDescription
	existingCourse.ImageURL = courseData.ImageURL

	if err := s.store.Courses.Update(existingCourse); err != nil {
		return nil, fmt.Errorf("error al actualizar el curso: %w", err)
	}

	return existingCourse, nil
}

func (s *courseService) UpdateCoursePatch(courseID uint, data map[string]interface{}) (*models.Course, error) {
	if courseID == 0 {
		return nil, errors.New("course ID cannot be zero")
	}

	var course dto.UpdateCourseRequest
	if err := utils.MapToStructStrict(data, &course); err != nil {
		return nil, errors.New("datos inválidos: " + err.Error())
	}

	if err := s.store.Courses.Patch(courseID, data); err != nil {
		return nil, err
	}

	updated, err := s.store.Courses.Get(courseID)
	if err != nil {
		return nil, errors.New("curso no encontrado")
	}

	return updated, nil
}

func (s *courseService) DeleteCourse(id uint) error {
	if err := s.store.Courses.Delete(id); err != nil {
		return fmt.Errorf("error al eliminar el curso: %w", err)
	}
	return nil
}

func (s *courseService) GetAllCourses() ([]*models.Course, error) {
	courses, err := s.store.Courses.GetAll()
	if err != nil {
		return nil, fmt.Errorf("error al obtener los cursos: %w", err)
	}
	return courses, nil
}

func (s *courseService) GetCourseWithModules(id uint) (*models.Course, error) {
	// This would require a repository method to preload modules
	course, err := s.store.Courses.Get(id)
	if err != nil {
		return nil, fmt.Errorf("curso no encontrado: %w", err)
	}

	// For now, return the course - would need to implement preloading in repository
	return course, nil
}

func (s *courseService) GetCoursesWithEnrollmentCount() ([]*models.Course, error) {
	// This would require a more complex repository method to calculate enrollment counts
	courses, err := s.store.Courses.GetAll()
	if err != nil {
		return nil, fmt.Errorf("error al obtener los cursos: %w", err)
	}
	return courses, nil
}

File: frontend/src/lib/controllers/course.ts
import { BaseController } from './base';
import type { Course, CreateCourseDTO, UpdateCourseDTO, Module, Enrollment } from '$lib/types';

export class CourseController extends BaseController {
	/**
	 * Get all courses
	 */
	async getCourses(): Promise<Course[]> {
		return this.get<Course[]>('/api/v1/courses');
	}

	/**
	 * Get a specific course by ID
	 */
	async getCourse(id: number): Promise<Course> {
		return this.get<Course>(`/api/v1/courses/${id}`);
	}

	/**
	 * Get course with all modules loaded
	 */
	async getCourseWithModules(id: number): Promise<Course> {
		return this.get<Course>(`/api/v1/courses/${id}/modules`);
	}

	/**
	 * Create a new course
	 */
	async createCourse(courseData: CreateCourseDTO): Promise<Course> {
		return this.post<Course>('/api/v1/courses', courseData);
	}

	/**
	 * Update an existing course
	 */
	async updateCourse(id: number, courseData: UpdateCourseDTO): Promise<Course> {
		return this.put<Course>(`/api/v1/courses/${id}`, courseData);
	}

	/**
	 * Partially update an existing course (PATCH)
	 */
	async updateCoursePatch(id: number, courseData: Partial<UpdateCourseDTO>): Promise<Course> {
		return this.patch<Course>(`/api/v1/courses/${id}`, courseData);
	}

	/**
	 * Delete a course
	 */
	async deleteCourse(id: number): Promise<void> {
		return this.delete(`/api/v1/courses/${id}`);
	}

	/**
	 * Get all modules for a specific course
	 */
	async getCourseModules(courseId: number): Promise<Module[]> {
		return this.get<Module[]>(`/api/v1/courses/${courseId}/modules`);
	}

	/**
	 * Get all enrollments for a specific course
	 */
	async getCourseEnrollments(courseId: number): Promise<Enrollment[]> {
		return this.get<Enrollment[]>(`/api/v1/courses/${courseId}/enrollments`);
	}
}
